BPA/Ch.10/js_super_indran.stan

```diff
@@ -0,0 +1,255 @@
+    for (k_rev in 0:(size(y_i) - 1)) {
+      int k;
+      k <- size(y_i) - k_rev;
+      if (y_i[k])
+        return k;
+    }
+    return 0;
+  }
+
+  /**
+   * Return a matrix of uncaptured probability
+   *
+   * @param nind        Number of individuals
+   * @param n_occasions Number of capture occasions
+   * @param p           Detection probability for each individual
+   *                    and capture occasion
+   * @param phi         Survival probability for each individual
+   *                    and capture occasion
+   *
+   * @return Uncaptured probability matrix
+   */
+  matrix prob_uncaptured(int nind, int n_occasions,
+                         matrix p, matrix phi) {
+    matrix[nind, n_occasions] chi;
+
+    for (i in 1:nind) {
+      chi[i, n_occasions] <- 1.0;
+      for (t in 1:(n_occasions - 1)) {
+        int t_curr;
+        int t_next;
+
+        t_curr <- n_occasions - t;
+        t_next <- t_curr + 1;
+        chi[i, t_curr] <- (1.0 - phi[i, t_curr]) +
+                          phi[i, t_curr] *
+                          (1.0 - p[i, t_next]) *
+                          chi[i, t_next];
+      }
+    }
+    return chi;
+  }
+}
+
+data {
+  int<lower=0> M;                              // Augmented sample size
+  int<lower=0> n_occasions;                    // Number of capture occasions
+  int<lower=0,upper=1> y[M, n_occasions];      // Augmented capture-history
+}
+
+transformed data {
+  int<lower=0,upper=n_occasions> first[M];
+  int<lower=0,upper=n_occasions> last[M];
+
+  for (i in 1:M)
+    first[i] <- first_capture(y[i]);
+  for (i in 1:M)
+    last[i] <- last_capture(y[i]);
+}
+
+parameters {
+  real<lower=0,upper=1> mean_phi;             // Mean survival
+  real<lower=0,upper=1> mean_p;               // Mean capture
+  real<lower=0,upper=1> psi;                  // Inclusion probability
+  vector<lower=0>[n_occasions] beta;
+  vector<lower=-15,upper=15>[M] epsilon;
+  real<lower=0,upper=5> sigma;
+}
+
+transformed parameters {
+  matrix<lower=0,upper=1>[M, n_occasions-1] phi;
+  matrix<lower=0,upper=1>[M, n_occasions] p;
+  simplex[n_occasions] b;                     // Entry probability
+  vector<lower=0,upper=1>[n_occasions] nu;
+  matrix<lower=0,upper=1>[M, n_occasions] chi;
+
+  // Constraints
+  for (i in 1:M) {
+    for (t in 1:(n_occasions - 1))
+      phi[i, t] <- mean_phi;
+    for (t in 1:n_occasions)
+      p[i, t] <- inv_logit(logit(mean_p) + epsilon[i]);
+  } //i
+
+  // Dirichlet prior for entry probabilities
+  // beta ~ gamma(1, 1);  // => model block
+  b <- beta / sum(beta);
+
+  // Convert entry probs to conditional entry probs
+  nu[1] <- b[1];
+  for (t in 2:(n_occasions - 1))
+    nu[t] <- b[t] / (1.0 - sum(b[1:(t - 1)]));
+  nu[n_occasions] <- 1.0;
+
+  // Uncaptured probability
+  chi <- prob_uncaptured(M, n_occasions, p, phi);
+}
+
+model {
+  vector[n_occasions] qnu;
+
+  qnu <- 1.0 - nu;
+
+  // Priors
+  mean_phi ~ uniform(0, 1);
+  mean_p ~ uniform(0, 1);
+  for (i in 1:M)
```

Comment:
This should be vectorized to

```
epsilon ~ normal(0, sigma);
```

And we don't like those interval constraints as you've declared for `epsilon` --- it'd be better to just leave it unconstrained; the hierarchical prior should control the scale.